README.md

```diff
@@ -814,6 +818,98 @@ a possible bug. See [the wiki page](https://github.com/WebAssembly/binaryen/wiki
+interchangeable, that is, they both lead to the same IR representation, so you
+can start with an annotated wat and have Binaryen write that to a binary + a
+source map file, or read a binary + source map file and print text which will
+contain those annotations.
+
+The IR representation of source map info is simple: in each function we have a
+map of expressions to their locations. Optimization passes should update the
+map as relevant. Often this "just works" because the optimizer tries to reuse
+nodes when possible, so they keep the same debug info.
+
+#### Shorthand notation
+
+The text format annotations support a shorthand in which repeated annotations
+are not necessary. For example, children are tagged with the debug info of the
+parent, if they have no annotation of their own:
+
+```wat
+;;@ src.cpp:100:33
+(i32.add
+  (i32.const 41)      ;; This receives an annotation of src.cpp:100:33
+  ;;@ src.cpp:111:44
+  (i32.const 1)
+)
+```
+
+The first const will have debug info identical to the parent, because it has
+none specified, and generally such nesting indicates a "bundle" of instructions
+that all implement the same source code.
+
+Note that text printing will not emit such repeated annotations, which can be
+confusing. To print out all the annotations, set `BINARYEN_PRINT_FULL=1` in the
+environment. That will print this for the above `add`:
+
+```wat
+[i32] ;;@ src.cpp:100:33
+(i32.add
+ [i32] ;;@ src.cpp:100:33
+ (i32.const 41)
+ [i32] ;;@ src.cpp:111:44
+ (i32.const 1)
+)
+```
+
+(full print mode also adds a `[type]` for each expression, right before the
+debug location).
+
+There is no shorthand in the binary format. That is, roundtripping (writing and
+reading) through a binary + source map should not change which expressions have
+debug info on them or the contents of that info.
+
+#### Implementation Details
+
+The [source maps format](https://sourcemaps.info/spec.html) defines a mapping
+using *segments*, that is, if a segment starts at binary offset 10 then it
+applies to all instructions at that offset and until another segment begins (or
+the end of the input is reached). Binaryen's IR represents a mapping from
+expressions to locations, as mentioned, so we need to map to and from the
+segment-based format when writing and reading source maps.
+
+That is mostly straightforward, but one thing we need to do is to handle the
+lack of debug info in between things that have it. If we have `A B C` where `B`
+lacks debug info, then just emitting a segment for `A` and `C` would lead `A`'s
+segment to also cover `B`, since in source maps segments do not have a size -
+rather they end when a new segment begins. To avoid `B` getting smeared in this
+manner, we emit a source maps entry to `B` of size 1, which just marks the
+binary offset it has, and without the later 3 fields of the source file, line
+number, and column. (This appears to be the intent of the source maps spec, and
+works in browsers and tools.)
```

Comment:
Source map allows one-character segments but I wasn't able to find what that meant while searching. Is this used for this meaning in JS too?

Reply:
Yes, AFAICT that's what the source maps spec intends for all users. The spec doesn't have much in terms of explanation or details (an example would have saved me much time...), but I think there is no other purpose for a 1-length segment, i.e. a segment with just a binary location and no debug info. It has to mean just that, to mark a location as not having debug info.

(This ambiguity would not have been an issue if segments had endings, but maybe that's less compact to encode...)

Reply:
> (This ambiguity would not have been an issue if segments had endings, but maybe that's less compact to encode...)

Sourcemap has line endings `;`, so I guess the inventors thought the line would end soon so it would have been mostly fine..? Which I'm not sure because minified JS doesn't have many newlines either. Also it doesn't apply to us because we only use a single line for binary...

Reply:
Interesting, yeah, maybe for wasm this is more of an issue since the binary is always a single "line".